### PR TITLE
Add support for blackboard types

### PR DESCRIPTION
**Pull request by bjoernschroeder**
Add type support for blackboard APIs. The types supported are bool, int, unsigned int, double, string & any (for any other type). This means that 
- pml files can use these types
- pml files can specify default values for them 
- APIs will throw meaningful exceptions when types mismatch in set & get API
- blackboard key mapping will also verify the types & throw if there are mismatches

Link to wrike task: https://www.wrike.com/open.htm?id=940589144

### CODE REVIEW

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 1]
```diff
@@ -23,25 +93,143 @@ class BlackboardImpl
...
+
     void clear() { vals.clear(); }
     bool empty() const { return vals.empty(); }
     size_t size() const { return vals.size(); }
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValueType> vals;
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/blackboard/Blackboard.h:182]: I think we really have a need to separate the blackboard from implementation at this point.  Ideally implementation with separate header file, that is included by `Blackboard.cpp` and not `Blackboard.h` (maybe pimpl?)

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 2]
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
...
+            convert<T>::get(vals.at(key));
+        } catch (const std::bad_variant_access& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        } catch (const std::bad_any_cast& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        }
     }
     template <typename T>
-    T& get(const std::string& key)
+    const T& get(const std::string& key)
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: why? This should return `T&`
And, there is no need to implement this method, instead can reuse the const method i.e.
`return const_cast<T&>(static_cast<const BlackboardImpl*>(this)->get<T>(key));`

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 3]
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
...
+            } else {
+                vals.emplace(key, value);
+            }
+        } else {
+            // value is predefined in pml file
+            std::string inputType = getNameFromType<T>();
+            std::string pmlType = getBlackboardValueType(key);
+            if (pmlType == "std::any") {
+                // insert as std::any with type T
+                vals.at(key) = std::any{value};
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: This will cause `get<T>` to fail if `T` is a known type (i.e. if `Find<T, BBValueType>::result` is true). So I guess easiest way is to check in `get<T>(key)` if pml type of key is std::any first & if so use `any_cast` there

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 4]
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
...
+                throw BlackboardTypeMismatch(inputType, pmlType);
+            }
+        }
+    }
+
+    void map(const std::string& srcKey, const std::string& targetKey, const BlackboardImpl& srcBb)
+    {
+        // Note: srcKey & targetKey has to be set
+        // mapping succeeds as long as targetType is constructible from srcType (so conversions are supported)
+        std::visit(
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h:163]: can you add a TODO here to eliminate compiler warnings? I haven't looked at how to do it yet

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 5]
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
...
     size_t size() const { return vals.size(); }
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValueType> vals;
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
+    YAML::Node getBlackboardValueNode(std::string key) const
+    {
+        for (const auto& entry : node) {
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: This is rather inefficient because it is called every time a key is set on the blackboard. Instead you could store the pml types of each key when the blackboard is created in an unordered_map

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 6]
```diff
@@ -111,13 +362,17 @@ class LockedBlackboardRW
...
+    const T& get(const std::string& key)
     {
         return _impl->get<T>(key);
     }
-    std::any& get(const std::string& key) { return _impl->get(key); }
+    BlackboardValueType& get(const std::string& key) { return _impl->get(key); }
 
-    void set(const std::string& key, const std::any& value) { _impl->set(key, value); }
+    template <typename T>
+    void set(const std::string& key, const T& value)
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: there is scope for improvement here. We should ideally have 3 ways to set values depending on how `T` is passed i.e. const lvalue, rvalue & emplace i.e.
`set(key, const T&`
`set(key, T&&`
`emplace(key, Args&&...)`

can do in another PR though because it will require forwarding `T` or `Args` through all the intermediate functions using `std::forward`

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 7]
```diff
@@ -23,25 +184,110 @@ class BlackboardImpl
...
+    template <class T>
+    void set(const std::string& key, const T& value)
+    {
+        if (getBlackboardValueNode(key).Type() == YAML::NodeType::Null) {
+            // value is not predefined in pml file
+            if (getNameFromType<T>() == "std::any") {
+                // if type is std::any, wrap into std::any
+                vals.emplace(key, std::any{value});
+            } else {
+                vals.emplace(key, value);
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: should not use `emplace` to set the value, because `emplace` fails if the key already exists on the blackboard

**Review by veerajsk — COMMENTED** [review]
need to review tests

File: alica_engine/include/engine/blackboard/BlackboardBlueprint.h [hunk 8]
```diff
@@ -17,10 +21,12 @@ class BlackboardBlueprint
     {
         vals.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(std::forward<decltype(args)>(args)...));
     }
+    void setBlackboardNode(const YAML::Node& node) { this->node = node; }
     friend class Blackboard;
 
 private:
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValueType> vals;
+    YAML::Node node;
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/BlackboardBlueprint.h]: The blueprint can store the object representation of the pml file (for blackboard section) i.e. an unordered_map of <key, type_name, std::optional<default_value>>. This way the YAML::Node need not be accessed/parsed by all internals of the engine

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 9]
```diff
@@ -1,15 +1,180 @@
...
 #include <unordered_map>
+#include <variant>
+#include <yaml-cpp/yaml.h>
+
+namespace YAML
+{
+template <>
+struct convert<std::monostate>
+{
+    static Node encode(const std::monostate& rhs)
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: can get rid of these methods & instead conditionally enable the compilation of the Parser i.e. using if constexpr

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 4 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h:163]: tried various ways to change the code to eliminate compiler warnings without success, will leave it as a TODO & suppress warnings for now

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 6 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h]: will be done in subsequent PR

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 1 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h:182]: most of the functions are templates but yeah, can look into how to do this. But can we defer to a separate PR?

**Comment by veerajsk** [comment]
@bjoernschroeder please review as well. I cannot add you as you are the creator of the PR

File: alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp [hunk 10]
```diff
@@ -216,12 +216,12 @@ AlicaDummyCommunication::AlicaDummyCommunication(const alica::AlicaCommunication
         : alica::IAlicaCommunication(callbacks)
         , _isRunning(false)
 {
-    s_modContainer.registerModule(this);
+    // s_modContainer.registerModule(this);
```
Comment (Dale-Koenig) [inline alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp]: What's going on here?

Comment (veerajsk) [reply alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp]: @bjoernschroeder I have the same question actually

Comment (Dale-Koenig) [reply alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp]: It does seem related to https://github.com/rapyuta-robotics/alica/pull/356/files

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 11]
```diff
@@ -1,47 +1,325 @@
...
+                }
+            }
+        }
     }
 
     template <typename T>
     const T& get(const std::string& key) const
     {
-        return std::any_cast<const T&>(vals.at(key));
+        // T must be an exact match to the T used while setting this key (i.e. no type conversions taken into account)
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: this comment is slightly inaccurate for keys that are not in the yaml file & whose type does not match exactly with any type in the variant. More precisely, T should be an exact match to the type stored in the variant.

File: supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp [hunk 12]
```diff
@@ -72,5 +75,47 @@ TEST_F(TestBlackBoard, testNotInheritBlackboardFlag)
     EXPECT_FALSE(alica::test::Util::getBasicPlan(ae, 1692837668719979457, 0)->getInheritBlackboard());
 }
 
+TEST_F(TestBlackBoard, testWithoutPlan)
```
Comment (veerajsk) [inline supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp]: need to add more tests

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 13]
```diff
@@ -1,47 +1,325 @@
...
+        }
     }
+
     template <typename T>
     T& get(const std::string& key)
     {
-        return std::any_cast<T&>(vals.at(key));
+        return const_cast<T&>(static_cast<const BlackboardImpl*>(this)->get<T>(key));
+    }
+    BBValueType& get(const std::string& key) { return _vals.at(key); }
```
Comment (veerajsk) [inline alica_engine/include/engine/blackboard/Blackboard.h]: can remove these?

File: alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp (hunk 10 above)
Comment (bjoernschroeder) [reply alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp]: Yes, during the implementation, the bug we had with the logger appeared more often, so I added the fix we discussed with abhi. It should be equal to your PR dale

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 13 above)
Comment (bjoernschroeder) [reply alica_engine/include/engine/blackboard/Blackboard.h]: These are not used anymore, so I dont see any reason to keep them.

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 14]
```diff
@@ -1,47 +1,325 @@
...
+private:
+    std::unordered_map<std::string, BBValueType> _vals;
+    std::unordered_map<std::string, std::string> _yamlType;
+
+    friend BlackboardUtil;
+
+    template <class T>
+    static const char* getTypeName()
+    {
+        auto typeIndex = BBValueType{T{}}.index() - 1;
```
Comment (bjoernschroeder) [inline alica_engine/include/engine/blackboard/Blackboard.h]: we can call `getTypeName(typeIndex)` here to avoid code duplication. Because getTypeName(std::size_t index) already accounts for monostate, you dont have to substract 1 from typeIndex then.

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 15]
```diff
@@ -1,47 +1,325 @@
...
+    {
+        template <class... Args>
+        static BBValueType make(std::size_t index, Args&&... args)
+        {
+            // make a BBValueType constructed from a value of type that is at `index` of the variant
+            // The value itself is either constructed from args or parsed from args
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
+            return makeHelper(index, std::make_index_sequence<BB_VALUE_TYPE_NAMES_SIZE + 2>(), std::forward<Args>(args)...);
```
Comment (bjoernschroeder) [inline alica_engine/include/engine/blackboard/Blackboard.h:262]: Can you add a comment why + 2 is used here? I think it was because we have to iterate over the 2 additional types in std::variant, monostate and any, right?

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 16]
```diff
@@ -1,47 +1,325 @@
...
+        return oss.str();
+    }
+
+public:
+    BlackboardImpl() = default;
+    BlackboardImpl(const BlackboardImpl&) = delete;
+    BlackboardImpl(const BlackboardBlueprint* blueprint)
+    {
+        for (auto it = blueprint->begin(); it != blueprint->end(); ++it) {
+            const auto& [key, keyInfo] = *it;
```
Comment (ntfshard) [inline alica_engine/include/engine/blackboard/Blackboard.h]: Feels like this 2 lines can be merged

File: alica_engine/src/engine/blackboard/BlackboardUtil.cpp [hunk 17]
```diff
@@ -14,24 +14,26 @@ void BlackboardUtil::setInput(const Blackboard* parent_bb, Blackboard* child_bb,
...
             Logging::logError("BlackboardUtil") << "Blackboard error passing " << parentKey << " into " << childKey << ". " << e.what();
+            std::rethrow_exception(std::current_exception());
         }
     }
 }
 
 void BlackboardUtil::setOutput(Blackboard* parent_bb, const Blackboard* child_bb, const KeyMapping* keyMapping)
 {
-    auto lockedParentBb = LockedBlackboardRW(*parent_bb);
+    const auto lockedParentBb = LockedBlackboardRW(*parent_bb);
```
Comment (ntfshard) [inline alica_engine/src/engine/blackboard/BlackboardUtil.cpp]: by-copy?

File: alica_engine/include/engine/blackboard/BlackboardBlueprint.h [hunk 18]
```diff
@@ -1,26 +1,32 @@
...
+
 public:
-    template <class... Args>
-    void registerValue(const std::string& key, Args&&... args)
+    using const_iterator = std::unordered_map<std::string, KeyInfo>::const_iterator;
+
+    void addKey(const std::string& key, const std::string& type, std::optional<std::string> defaultValue)
     {
-        vals.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(std::forward<decltype(args)>(args)...));
+        _keyInfo.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(KeyInfo{.type = type, .defaultValue = defaultValue}));
```
Comment (ntfshard) [inline alica_engine/include/engine/blackboard/BlackboardBlueprint.h]: `{.type = type` technically requires c++20

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 19]
```diff
@@ -104,20 +386,22 @@ class LockedBlackboardRW
...
     {
-        _impl->registerValue(key, std::forward<Args>(args)...);
+        return _impl->get<T>(key);
     }
-
     template <typename T>
     T& get(const std::string& key)
     {
         return _impl->get<T>(key);
     }
```
Comment (ntfshard) [inline alica_engine/include/engine/blackboard/Blackboard.h]: technically one can be implemented, and other could be made with const_cast. But I guess this is also OK

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 15 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h:262]: yes

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 19 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h]: actually, it is better this way since this is just a wrapper class, so it should just forward the call

File: alica_engine/src/engine/blackboard/BlackboardUtil.cpp (hunk 17 above)
Comment (veerajsk) [reply alica_engine/src/engine/blackboard/BlackboardUtil.cpp]: changed the syntax

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 20]
```diff
@@ -1,47 +1,330 @@
...
+    struct makeBBValueForIndex
+    {
+        template <class... Args>
+        static BBValueType make(std::size_t index, Args&&... args)
+        {
+            // make a BBValueType constructed from a value of type that is at `index` of the variant
+            // The value itself is either constructed from args or parsed from args
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
```
Comment (ntfshard) [inline alica_engine/include/engine/blackboard/Blackboard.h]: index is unsigned type. It can be !=0 and/or can be less some upper bound

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 21]
```diff
@@ -1,47 +1,330 @@
...
+            if (BB_VALUE_TYPE_NAMES[i] == type) {
+                // i + 1 to account for std::monostate in the variant
+                return i + 1;
+            }
+        }
+        return {};
+    }
+
+    static constexpr const char* BB_VALUE_TYPE_NAMES[] = {"bool", "int64", "double", "std::string"};
+    static constexpr std::size_t BB_VALUE_TYPE_NAMES_SIZE = sizeof(BB_VALUE_TYPE_NAMES) / sizeof(const char*);
```
Comment (ntfshard) [inline alica_engine/include/engine/blackboard/Blackboard.h:248]: std::array?

File: supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp [hunk 22]
```diff
@@ -72,5 +118,359 @@ TEST_F(TestBlackBoard, testNotInheritBlackboardFlag)
...
+    // unknown types
+    bbrw.set<PlanStatus>("val5", PlanStatus::Success);
+    EXPECT_EQ(bbrw.get<PlanStatus>("val5"), PlanStatus::Success);
+}
+
+TEST_F(TestBlackboard, testMappingFromBool)
+{
+    EXPECT_TRUE((checkMapping<bool, bool>(BBType::BOOL, true, BBType::BOOL, true)));
+    EXPECT_TRUE((checkMapping<bool, int64_t>(BBType::BOOL, true, BBType::INT64, 1)));
+    EXPECT_TRUE((checkMapping<bool, double>(BBType::BOOL, true, BBType::DOUBLE, 1.0)));
```
Comment (Dale-Koenig) [inline supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp]: I don't think bool to double (or even int) conversations are something we really want to support

Comment (veerajsk) [reply supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp]: A single line change in 292 will determine the kind of conversions we support. Right now I use `std::is_constructible`. But we could of course use our own custom trait if we want specific conversions to be allowed/disallowed.

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 23]
```diff
@@ -1,47 +1,322 @@
...
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
+            // Note: make index sequence equal to the number of types in BBValueType
+            return makeHelper(index, std::make_index_sequence<BB_VALUE_TYPE_NAMES_SIZE + 2>(), std::forward<Args>(args)...);
+        }
+
+    private:
+        template <class T>
+        struct Parser
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/blackboard/Blackboard.h:267]: Can be separate PR, but IMO would be better if we can remove the yaml parsing from blackboard and preprocess it.

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 24]
```diff
@@ -1,47 +1,322 @@
...
+            {
+                YAML::Node node = YAML::Load(value);
+                return node.as<T>();
+            }
+        };
+
+        template <class... Args, std::size_t... Is>
+        static BBValueType makeHelper(std::size_t index, std::index_sequence<Is...>, Args&&... args)
+        {
+            BBValueType vals[] = {makeBBValueIfIndex<Is>::make(index, std::forward<Args>(args)...)...};
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/blackboard/Blackboard.h]: std::array

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 25]
```diff
@@ -1,47 +1,322 @@
...
+
+        template <std::size_t INDEX>
+        struct makeBBValueIfIndex
+        {
+            using TypeAtIndex = std::decay_t<decltype(std::get<INDEX>(std::declval<BBValueType>()))>;
+
+            template <class... Args>
+            static BBValueType make(std::size_t index, Args&&... args)
+            {
+                // if INDEX == index, make a variant with a value of type that is same as the variant's type at index INDEX, intialized (or parsed) with value
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/blackboard/Blackboard.h]:
```suggestion
                // if INDEX == index, make a variant with a value of type that is same as the variant's type at index INDEX, initialized (or parsed) with value
```

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 26]
```diff
@@ -1,47 +1,322 @@
...
+                // else makes a invalid variant, i.e. with value monostate
+                if constexpr (PARSE_ARGS) {
+                    if constexpr (!std::is_same_v<TypeAtIndex, std::monostate> && !std::is_same_v<TypeAtIndex, std::any>) {
+                        return index == INDEX ? BBValueType{TypeAtIndex{Parser<TypeAtIndex>{}(args)...}} : BBValueType{};
+                    }
+                } else {
+// TODO: figure out a way to disable warnings, caused due to std::visit. Note: really tricky since small changes can cause unexpected behaviour, so best to
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/blackboard/Blackboard.h]: Ideally we should not allow narrowing conversions:  i.e. do not allow mapping float to int.  In fact I think the only conversion worth maybe supporting is int -> double

File: supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp (hunk 22 above)
Comment (Dale-Koenig) [reply supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp]: As mentioned in comment above, I'm not convinced there is value supporting any conversion other than int->double.  Especially narrowing conversions should not be supported imo

Comment (veerajsk) [reply supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp]: Yeah, I did look for a way to prevent narrowing conversions (that also causes the compiler warnings which had to be suppressed) but it turned out to be quite tricky. 
Also bool -> double is not a narrowing conversion so maybe it needs to be dealt with separately.
Let me look into this a bit & see if I can come up with a suitable trait

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 23 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h:267]: Could be tricky to do because the types are inferred here. Probably will need sharing all the code which does the type inference between the blackboard & the blueprint.

Comment (Dale-Koenig) [reply alica_engine/include/engine/blackboard/Blackboard.h:267]: Could also just disable all conversions at first and enable later if we figure it out

File: supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp (hunk 22 above)
Comment (veerajsk) [reply supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp]: I found a solution. Basically I now call the set API here to do the mapping. This implies that mapping & set now support the same kind of type conversions which is desirable. The only type conversions supported are promotions i.e. promotions from smaller integral types to 64 bit ones

**Review by Dale-Koenig — APPROVED** [review]

**Comment by Dale-Koenig** [comment]
PR description, link wrike task, etc

File: alica_engine/include/engine/blackboard/Blackboard.h [hunk 27]
```diff
@@ -1,47 +1,331 @@
...
-        vals.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(std::forward<decltype(args)>(args)...));
+        std::ostringstream oss;
+        (oss << ... << std::forward<Args>(args));
+        return oss.str();
+    }
+
+public:
+    BlackboardImpl() = default;
+    BlackboardImpl(const BlackboardImpl&) = delete;
+    BlackboardImpl(const BlackboardBlueprint* blueprint)
```
Comment (Dale-Koenig) [inline alica_engine/include/engine/blackboard/Blackboard.h:50]: I think parsing the yaml etc should be done in the blueprint, not when constructing the blackboard

**Review by veerajsk — APPROVED** [review]

File: alica_engine/include/engine/blackboard/Blackboard.h (hunk 27 above)
Comment (veerajsk) [reply alica_engine/include/engine/blackboard/Blackboard.h:50]: ok, I will create another task for this & see how it can be done because the parsing related code is currently tied up in the blackboard class. Wrike task: https://www.wrike.com/open.htm?id=1033598893
